### PR TITLE
Encode/decode enums as well when they're at the roots of JSON messages

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    out.value(enumJsonFormatter.toStringOrNumber(value))
+  }
+
+  @Throws(IOException::class)
+  override fun fromJson(input: JsonReader): E {
+    val path: String = input.path
+    val nextString = input.nextString()
+
+    return enumJsonFormatter.fromString(nextString)
+        ?: throw JsonDataException("Unexpected $nextString at path $path")
```
Comment (oldergod) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt:43]: Is it too simple?

Comment (swankjesse) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt:43]: For followup, lets use `Options` for these.

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.IOException
+import java.util.Arrays
+
+internal class EnumJsonAdapter<E>(
+  private val enumJsonFormatter: EnumJsonFormatter<E>,
+) : JsonAdapter<E>() where E : Enum<E>, E : WireEnum {
+
+  @Throws(IOException::class)
+  override fun toJson(out: JsonWriter, value: E?) {
+    if (value == null) return
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt:32]: no no no no, this is definitely broken. You gotta do `out.nullValue()`

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    out.value(enumJsonFormatter.toStringOrNumber(value))
+  }
+
+  @Throws(IOException::class)
+  override fun fromJson(input: JsonReader): E {
+    val path: String = input.path
```
Comment (swankjesse) [inline wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt:39]: don’t compute the path unless you need it; that’ll wreck performance

**Review by swankjesse — APPROVED** [review]

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt (hunk 2 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt:32]: 4 no's

File: wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt (hunk 3 above)
Comment (oldergod) [reply wire-library/wire-moshi-adapter/src/main/java/com/squareup/wire/EnumJsonAdapter.kt]: inlined

**Pull request merged**